### DISCUSSION TITLE
New Maven Bazel Support Tries to Update Stable to Unstable

### DISCUSSION THREAD

**Discussion by cpsauer** [renovatebot/renovate#20396] [Suggest an Idea]
Hey, wonderful Renovate folks,

I really appreciate the Bazel improvements you've been landing recently--and just got my first Maven PR from https://github.com/renovatebot/renovate/pull/20351

There's one snag, though: It's trying to update a stable release to an unstable one.
I *think* that might be a bug, since Renovate (by default) has always seemed to have the desired behavior of updating stable dependencies to newer stable versions only, but I wanted to check in first since the bug report issue template said confirmed bugs only.

For more concrete details:
Here's the log showing the update from 1.3.0 (stable) to 1.6.0-dev01 (unstable, based on the version name). 
```
          {
            "datasource": "maven",
            "depName": "androidx.annotation:annotation",
            "currentValue": "1.3.0",
            "depType": "maven_install",
            "registryUrls": [
              "https://maven.google.com",
              "https://repo1.maven.org/maven2"
            ],
            "managerData": {
              "idx": 15
            },
            "depIndex": 15,
            "updates": [
              {
                "bucket": "latest",
                "newVersion": "1.6.0-dev01",
                "newValue": "1.6.0-dev01",
                "releaseTimestamp": "2023-02-08T18:00:00.000Z",
                "newMajor": 1,
                "newMinor": 6,
                "updateType": "minor",
                "branchName": "renovate/androidx.annotation-annotation-1.x"
              }
            ],
            "warnings": [],
            "versioning": "maven",
            "sourceUrl": "https://cs.android.com/androidx/platform/frameworks/support",
            "registryUrl": "https://maven.google.com",
            "homepage": "https://developer.android.com/jetpack/androidx/releases/annotation#1.6.0-dev01",
            "currentVersion": "1.3.0",
            "isSingleVersion": true,
            "fixedVersion": "1.3.0"
          },
```

Thanks so much,
Chris

[Tagging @zharinov, since he was the author of https://github.com/renovatebot/renovate/pull/20351, but feel free anyone to chime in if you have thoughts.]

**Comment by rarkins** [comment]
I think the answer here is the same as https://github.com/renovatebot/renovate/discussions/15828#discussioncomment-2863516

i.e. `-whatever` suffixes don't always mean "unstable" in Maven, like they do in semver. Just because it says `-DO-NOT-USE-REALLY-UNSTABLE` does not make that meaningful to a machine.

Pragmatically, @zharinov has added some suffixes to our list which aren't technically considered unstable (they sort HIGHER than the unsuffixed version) but which we override to make unstable. `-dev` may be another one we can add.

**Reply by zharinov** [reply]
This eventually will put us in the position when we diverge substantially from the standard. I'd rather would think about versioning schemes pre-defined for particular well-known packages/groups, because it seems to be what some package creators are doing: they use other versioning schemes for their Maven packages.

**Reply by rarkins** [reply]
I don't want to be too strict about it though. `-dev` seems like an example which is almost certainly worth it to treat as unstable

**Reply by zharinov** [reply]
https://github.com/renovatebot/renovate/blob/9eb8684ba54c73e7319ed68d70e53547ff42fa00/lib/modules/versioning/maven/compare.ts#L169-L177

Where does it fit better?

**Reply by rarkins** [reply]
I think before alpha actually

**Reply by cpsauer** [reply]
Thanks so much for taking the time to think about this and reply, guys. Really appreciate your thoughtfulness and additional context.
[And sorry I didn't see and reference before. Thanks for the crossref and all the learnings.]

Equipped now with better understanding, dev does seem like a tricky decision for all the reasons you both said. One one hand, yeah, it should almost always be unstable. On the other--now that I understand better--Google really shouldn't be using it. It's [breaking their own release page](https://developer.android.com/jetpack/androidx/versions), and annotation seems like the only one currently broken this way--even if popular one. FWIW, [here they mean for dev to be after beta but before rc](https://developer.android.com/jetpack/androidx/releases/annotation#1.6.0-dev01), but I agree that that doesn't make a lot of sense.

I'm finding myself wondering if another, more general solution might be to not update to a new custom modifier unless already on that modifier, but I bet you all have additional thoughts and context?

For now, I'll workaround google's weirdness with `packageRules` -> `"versioning": "gradle"` (updated per below, parallel to what you suggest in the other discussion).

Thanks again!
Chris

**Reply by cpsauer** [reply]
Hey, guys! Coming back in with an update after discussing with some folks at Google about the release page and other issues.

Sounds like they're assuming *gradle* versioning, and likely that's at the root of the issues, including their release page. [Gradle does sort dev lowest among non-numeric parts, and treats all as lower than rc, meaning they're all unstable](https://docs.gradle.org/current/userguide/single_versions.html).

So I've changed the workaround to assume things are gradle-versioned for Bazel maven. 
It strikes me that gradle versioning *might* make sense as a general default for Bazel's maven rules, since Bazel is Google and the sorting rules are similar enough that it might be a better default on average? That might at least be better than special casing suffixes? It looks also like they have a [PR coming soon that's switching everything to Gradle resolution by default](https://github.com/bazelbuild/rules_jvm_external/pull/807) (though I'll check in to confirm). [A more sophisticated approach might be to condition on the use-gradle markers in the POM files, but I'm assuming that'd be much more work.]

**Reply by rarkins** [reply]
Looks like we already cover that for gradle versioning: https://github.com/renovatebot/renovate/blob/56c161a8d4bc8ebf661a641fdcd12d803492749f/lib/modules/versioning/gradle/index.ts#L76-L89

Defaulting _maven_ dependencies in Bazel to use _gradle_ versioning seems strange though, but I don't object if that's how Google' gone and done it

**Reply by zharinov** [reply]
But in documentation, we may want to provide the example configuration of how to set Maven versioning instead

**Reply by viceice** [reply]
maybe we should add a docs note to the bazel manager for now?

**Reply by cpsauer** [reply]
After letting it simmer for a couple days, changing the default to gradle and documenting still sounds like the right solution to me--given the unfortunate reality of one datasource, maven, accessed with two different semantics, gradle and maven. I think that's especially true given Google, the tool owner, is more focused around gradle (and happy to break Maven-only setups with this new kotlin Multiplatform push), and since I'd expect the vast majority of packages to work with Gradle.

In case useful for the docs--and as a double check--here's the config snippet I'd been using, modified to reset things to maven.

```json
  "packageRules": [{
    "matchManagers": ["bazel"],
    "matchDatasources": ["maven"],
    "versioning": "maven"
  }],
```

**Reply by cpsauer** [reply]
Thanks so much guys, as always, for being so wonderfully thoughtful and responsive.

**Reply by rarkins** [reply]
@cpsauer I'm a little confused by your last example. The maven datasource defaults to maven versioning, so why is your package rule necessary? I thought based on the above discussion that you'd be overriding it to versioning=gradle

**Reply by cpsauer** [reply]
Yes indeed. In my current config it instead sets versioning to gradle; that one I posted is modified (potentially for the docs) as an example of resetting things back to maven, if you guys decide to change the default to gradle.

**Reply by rarkins** [reply]
Gotcha, we should add that to the bazel docs if we make this change

**Reply by rarkins** [reply]
@cpsauer could you raise a new Feature Request issue for this now? The description can be short and link to this Discussion - I think it's as simple as "Use gradle versioning for maven dependencies found by bazel"? No need for a reproduction either unless you wish to

**Reply by cpsauer** [reply]
Sure thing. Gimme a sec.
Thanks for the encouragement to keep it short (and fast).

**Reply by cpsauer** [reply]
Donezo https://github.com/renovatebot/renovate/issues/20725

**Reply by zharinov** [reply]
We've done it, you may want to check the "retry" box in your PRs of interest to update it.

**Reply by cpsauer** [reply]
Thanks so much, again, guys. I'm sure this'll save a bunch of people configuration time using the new maven bazel support.
(Backing out my workaround presently.)
